src/replay/replay_recorder.cpp

```diff
@@ -116,13 +129,125 @@ void ReplayRecorder::update(int ticks)
+        }
+        if (powerup_type == -1)
+        {
+            Log::error("ReplayRecorder", "Unknown powerup type");
+            return;
+        }
+
+        if (m_count_transforms[i] >= 2)
+        {
+            BonusInfo *b_prev       = &(m_bonus_info[i][m_count_transforms[i]-1]);
+            BonusInfo *b_prev2      = &(m_bonus_info[i][m_count_transforms[i]-2]);
+            PhysicInfo *q_prev      = &(m_physic_info[i][m_count_transforms[i]-1]);
+
+            // If the kart changes its steering
+            if (fabsf(kart->getControls().getSteer() - m_previous_steer) >
+                                            stk_config->m_replay_delta_steering)
+                force_update = true;
+
+            // If the kart starts or stops skidding
+            if (kart->getSkidding()->getSkidState() != q_prev->m_skidding_state)
+                force_update = true;
+            // If the kart changes speed significantly
+            float speed_change = fabsf(kart->getSpeed() - q_prev->m_speed);
+            if ( speed_change > stk_config->m_replay_delta_speed )
+            {
+                if (speed_change > 4*stk_config->m_replay_delta_speed)
+                    force_update = true;
+                else if (speed_change > 2*stk_config->m_replay_delta_speed &&
+                         time - m_last_saved_time[i] > (stk_config->m_replay_dt/8.0f))
+                    force_update = true;
+                else if (time - m_last_saved_time[i] > (stk_config->m_replay_dt/3.0f))
+                    force_update = true;
+            }
+
+            // If the attachment has changed
+            if (attachment != b_prev->m_attachment)
+                  force_update = true;
+    
+            // If the item amount has changed
+            if (kart->getNumPowerup() != b_prev->m_item_amount)
+                force_update = true;
+
+            // If the item type has changed
+            if (powerup_type != b_prev->m_item_type)
+                force_update = true;
+
+            // In egg-hunt mode, if an egg has been collected
+            // In battle mode, if a live has been lost/gained
+            if (special_value != b_prev->m_special_value)
+                force_update = true;
+
+            // If nitro starts being used or is collected
+            if (kart->getEnergy() != b_prev->m_nitro_amount &&
+                b_prev->m_nitro_amount == b_prev2->m_nitro_amount)
+                force_update = true;
+
+            // If nitro stops being used
+            // (also generates an extra transform on collection,
+            //  should be negligible and better than heavier checks)
+            if (kart->getEnergy() == b_prev->m_nitro_amount &&
+                b_prev->m_nitro_amount != b_prev2->m_nitro_amount)
+                force_update = true;
+
+            // If close to the end of the race, reduce the time step
+            // for extra precision
+            // TODO : fast updates when close to the last egg in egg hunt
```

Comment:
A left-over TODO?

Reply:
This one is actually not a left-over : it would need code to compute the kart's distance with still available egg items. This is doable, but a bit complex for a small benefit, so I didn't want to stall the PR for it.